lib/extensions.js

```diff
@@ -111,37 +111,39 @@ module.exports = function(realmConstructor, context) {
     setConstructorOnPrototype(realmConstructor.Results);
     setConstructorOnPrototype(realmConstructor.Object);
 
-    Object.defineProperty(realmConstructor.Object.prototype, "toJSON", {
-        value: function () {
-            const result = {}
-            Object.keys(this).forEach(p => result[p] = this[p]);
-            Object.keys(Object.getPrototypeOf(this)).forEach(p => result[p] = this[p]);
-            return result;
+    Object.defineProperty(realmConstructor.Collection.prototype, "toJSON", {
+        value: function (key, cache = new Map()) {
+            return this.map(item => item.toJSON(key, cache));
         },
 
         writable: true,
         configurable: true,
         enumerable: false
     });
 
-    Object.defineProperty(realmConstructor.Object.prototype, "keys", {
-        value: function () {
-            return Object.keys(this).concat(Object.keys(Object.getPrototypeOf(this)));
-        },
-
-        writable: true,
-        configurable: true,
-        enumerable: false
-    });
+    Object.defineProperty(realmConstructor.Object.prototype, "toJSON", {
+        value: function (key, cache = new Map()) {
+            const id = this._objectId();
 
-    Object.defineProperty(realmConstructor.Object.prototype, "entries", {
-        value: function () {
-            let result = {};
-            for (const key in this) {
-                result[key] = this[key];
+            // check if currect objectId has already matched
+            const existing = cache.get(id);
+            if (existing) {
+                return existing;
             }
 
-            return Object.entries(result);
+            // create new result, and store in cache
+            const result = { $realmId: id };
+            cache.set(id, result);
+
+            const moveAsSerializable = (cache, key, from, to) => {
+                // TODO: check that we're dealing with a Realm type, before passing in cache?
+                to[key] = from[key].toJSON ? from[key].toJSON(undefined, cache) : this[key];
+            };
+
+            Object.keys(this).forEach(k => moveAsSerializable(cache, k, this, result));
+            Object.keys(Object.getPrototypeOf(this)).forEach(k => moveAsSerializable(cache, k, this, result));
```

Comment:
Probably nit picking here, but it would make it a bit more DRY.
```suggestion
            Object.keys(this)
                .concat(Object.keys(Object.getPrototypeOf(this)));
                .forEach(k => moveAsSerializable(cache, k, this, result));
```

Reply:
I like it. It also opens for having `moveAsSerializable` back in as an anonymous function.